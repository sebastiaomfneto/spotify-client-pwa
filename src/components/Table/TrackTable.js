import React from "react";

import { Table } from "./Table";
import { FavoriteButton } from "../FavoriteButton";

import { buildImageSrc } from "../../helpers";

function buildArtist(artists = []) {
  return artists.map((a) => a.name).join(", ");
}

function buildDuration(ms = 0) {
  const min = (ms / 1000 / 60) << 0;
  let sec = (ms / 1000) % 60 << 0;

  if (sec < 10) {
    sec = sec.toString() + "0";
  }

  return `${min}:${sec}`;
}

export function TrackTable({ album, tracks = [] }) {
  return (
    <Table columns={["", "Album Image", "Name", "Artist", "Album", "Duration"]}>
      {tracks.map((track) => (
        <tr key={track.id}>
          <td>
            <FavoriteButton id={track.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(track.album?.images ?? album.images)}
              alt="track"
              width="20px"
              height="20px"
            />
          </td>
          <td>{track.name}</td>
          <td>{buildArtist(track.artists)}</td>
          <td>{track.album?.name ?? album.name}</td>
          <td>{buildDuration(track.duration_ms)}</td>
        </tr>
      ))}
    </Table>
  );
}
