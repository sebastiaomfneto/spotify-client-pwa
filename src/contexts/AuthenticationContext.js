import React, { useState, useEffect } from "react";

import { getItem, setItem } from "../services/StorageService";

export const AuthenticationContext = React.createContext();

export function AuthenticationContextProvider({ children }) {
  const initialToken = getItem("token");
  const [token, setToken] = useState(initialToken);
  const [isAuthenticated, setIsAuthenticated] = useState(!!initialToken);

  useEffect(() => {
    setItem("token", token);
    setIsAuthenticated(!!token);
  }, [token]);

  return (
    <AuthenticationContext.Provider
      value={{
        token,
        isAuthenticated,
        setToken,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
