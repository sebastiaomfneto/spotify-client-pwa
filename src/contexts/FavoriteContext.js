import React, { useEffect, useReducer } from 'react';

import * as StorageService from '../services/StorageService';
import { NotImplementedError } from '../helpers';

export const FavoriteContext = React.createContext();

export const TOGGLE_FAVORITE = 'TOGGLE_FAVORITE';

function initializer(initialState) {
  return StorageService.getItem('favorites') ?? initialState;
}

function reducer(state, action) {
  switch (action.type) {
    case TOGGLE_FAVORITE:
      const isfavorited = state.some((f) => f === action.payload);
      if (isfavorited) {
        return state.filter((favorite) => favorite !== action.payload);
      }
      return state.concat(action.payload);
    default:
      throw new NotImplementedError();
  }
}

export function FavoriteContextProvider({ initialState = [], children }) {
  const [state, dispatch] = useReducer(reducer, initialState, initializer);

  useEffect(() => {
    StorageService.setItem('favorites', state);
  }, [state]);

  return (
    <FavoriteContext.Provider
      value={{
        state,
        dispatch,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
}
