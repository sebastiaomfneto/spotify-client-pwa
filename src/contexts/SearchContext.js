import React, { useReducer } from 'react';

import { NotImplementedError } from '../helpers';

export const SearchContext = React.createContext();

export const SET_SEARCH = 'SET_SEARCH';

const initialState = '';

function reducer(_state, action) {
  switch (action.type) {
    case SET_SEARCH:
      return action.payload;
    default:
      throw new NotImplementedError();
  }
}

export function SearchContextProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <SearchContext.Provider
      value={{
        state,
        dispatch,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
}
