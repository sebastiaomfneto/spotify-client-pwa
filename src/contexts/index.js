export * from './AuthenticationContext';
export * from './FavoriteContext';
export * from './SearchContext';
