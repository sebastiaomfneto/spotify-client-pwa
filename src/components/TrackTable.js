import React from "react";

import Table from "../components/Table";
import FavoriteButton from "../components/FavoriteButton";

import { buildImageSrc } from "../helpers";

export function buildArtist(artists = []) {
  return artists.map((a) => a.name).join(", ");
}

export function buildDuration(ms = 0) {
  const min = (ms / 1000 / 60) << 0;
  let sec = (ms / 1000) % 60 << 0;

  if (sec < 10) {
    sec = sec.toString() + "0";
  }

  return `${min}:${sec}`;
}

export default function TrackTable({ album, tracks = [] }) {
  return (
    <Table columns={["", "Album Image", "Name", "Artist", "Album", "Duration"]}>
      {tracks.map((track) => (
        <tr key={track.id}>
          <td>
            <FavoriteButton id={track.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(track.album?.images ?? album.images)}
              alt="track"
              width="20px"
              height="20px"
            />
          </td>
          <td>{track.name}</td>
          <td>{buildArtist(track.artists)}</td>
          <td>{track.album?.name ?? album.name}</td>
          <td>{buildDuration(track.duration_ms)}</td>
        </tr>
      ))}
    </Table>
  );
}
