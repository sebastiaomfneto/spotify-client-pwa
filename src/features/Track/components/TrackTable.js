import React from 'react';

import { Table, FavoriteButton } from '../../../components';
import { buildImageSrc } from '../../../helpers';
import { buildArtist, buildDuration } from '../helpers';

export function TrackTable({ album, tracks = [] }) {
  return (
    <Table columns={['', 'Album Image', 'Name', 'Artist', 'Album', 'Duration']}>
      {tracks.map((track) => (
        <tr key={track.id}>
          <td>
            <FavoriteButton id={track.id} />
          </td>
          <td>
            <img
              src={buildImageSrc(track.album?.images ?? album.images)}
              alt="track"
              width="20px"
              height="20px"
            />
          </td>
          <td>{track.name}</td>
          <td>{buildArtist(track.artists)}</td>
          <td>{track.album?.name ?? album.name}</td>
          <td>{buildDuration(track.duration_ms)}</td>
        </tr>
      ))}
    </Table>
  );
}
