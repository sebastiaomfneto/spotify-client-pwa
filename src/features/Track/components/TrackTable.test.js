import React from 'react';
import { render } from '@testing-library/react';

import { FavoriteContext } from '../../../contexts/FavoriteContext';
import { TrackTable } from './TrackTable';

jest.mock('../../../services/StorageService');

const withProviders = (element, state = [], dispatch) => (
  <FavoriteContext.Provider value={{ state, dispatch }}>
    {element}
  </FavoriteContext.Provider>
);

describe('testing TrackTable', () => {
  const album = {
    name: 'TEST_ALBUM',
    images: [{ src: 'TEST_SRC' }],
  };

  const tracks = [
    {
      id: 1,
      album,
      artists: [{ name: 'TEST_ARTIST_1' }, { name: 'TEST_ARTIST_1' }],
      duration_ms: 100,
    },
  ];

  test('should match with snapshot', () => {
    const { container } = render(
      withProviders(<TrackTable album={album} tracks={tracks} />)
    );

    expect(container).toMatchSnapshot();
  });
});
