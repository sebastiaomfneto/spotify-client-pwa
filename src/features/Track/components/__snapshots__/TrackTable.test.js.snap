// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing TrackTable should match with snapshot 1`] = `
<div>
  <div
    class="Table"
  >
    <table>
      <thead>
        <tr>
          <th />
          <th>
            Album Image
          </th>
          <th>
            Name
          </th>
          <th>
            Artist
          </th>
          <th>
            Album
          </th>
          <th>
            Duration
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <button
              class="FavoriteButton "
              data-testid="FavoriteButton-button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-heart fa-w-16 "
                data-icon="heart"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
          <td>
            <img
              alt="track"
              height="20px"
              width="20px"
            />
          </td>
          <td />
          <td>
            TEST_ARTIST_1, TEST_ARTIST_1
          </td>
          <td>
            TEST_ALBUM
          </td>
          <td>
            00:00
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
